File: Database/StudentCodeDataStore.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.SqlClient;
using System.Threading.Tasks;
using DevSpace.Common;

namespace DevSpace.Database {
	public class StudentCodeDataStore : IDataStore<IStudentCode> {
		public async Task<IStudentCode> Get( int Id ) {
			using( SqlConnection connection = new SqlConnection( Settings.ConnectionString ) ) {
				connection.Open();

				using( SqlCommand command = new SqlCommand( "SELECT * FROM StudentCodes WHERE Id = @Id", connection ) ) {
					command.Parameters.Add( "Id", SqlDbType.Int ).Value = Id;

					using( SqlDataReader dataReader = await command.ExecuteReaderAsync() ) {
						if( await dataReader.ReadAsync() ) {
							return new Models.StudentCodeModel( dataReader );
						}
					}
				}
			}

			return null;
		}

		public async Task<IList<IStudentCode>> GetAll() {
			throw new NotImplementedException();
		}

		public async Task<IList<IStudentCode>> Get( string Field, object Value ) {
			List<IStudentCode> returnList = new List<IStudentCode>();

			using( SqlConnection connection = new SqlConnection( Settings.ConnectionString ) ) {
				connection.Open();

				using( SqlCommand command = new SqlCommand( string.Format( "SELECT * FROM StudentCodes WHERE {0} = @value", Field ), connection ) ) {
					command.Parameters.AddWithValue( "value", Value );

					using( SqlDataReader dataReader = await command.ExecuteReaderAsync() ) {
						while( await dataReader.ReadAsync() ) {
							returnList.Add( new Models.StudentCodeModel( dataReader ) );
						}
					}
				}
			}

			return returnList;
		}

		public async Task<IStudentCode> Add( IStudentCode ItemToAdd ) {
			if( string.IsNullOrWhiteSpace( ItemToAdd.Code ) ) return null;
			if( string.IsNullOrWhiteSpace( ItemToAdd.Email ) ) return null;

			Models.StudentCodeModel returnValue = new Models.StudentCodeModel {
				Email = ItemToAdd.Email,
				Code = ItemToAdd.Code
			};

			using( SqlConnection connection = new SqlConnection( Settings.ConnectionString ) ) {
				connection.Open();

				using( SqlCommand command = new SqlCommand( "INSERT StudentCodes ( Email, Code ) VALUES ( @Email, @Code ); SELECT SCOPE_IDENTITY();", connection ) ) {
					command.Parameters.Add( "Code", SqlDbType.VarChar ).Value = ItemToAdd.Code;
					command.Parameters.Add( "Email", SqlDbType.VarChar ).Value = ItemToAdd.Email;
					returnValue.Id = Convert.ToInt32( await command.ExecuteScalarAsync() );
				}
			}

			return returnValue;
		}

		public Task<IStudentCode> Update( IStudentCode ItemToUpdate ) {
			throw new NotImplementedException();
		}
	}
}

File: FSharp.Database/StudentCodeDataStore.fs
namespace DevSpace.FSharp.Database

open Microsoft.FSharp.Control
open DevSpace.FSharp.Common
open System.Data
open System.Data.SqlClient

type StudentCodeDataStore() =
  inherit DataStore<StudentCode>()

  let connectionString = "Server=localhost;Database=DevSpace;Trusted_Connection=True;"

  let getStudentCodeFromDataReader (dataReader:SqlDataReader) : StudentCode = {
      Id    = dataReader.GetInt32 0;
      Email = dataReader.GetString 1;
      Code  = dataReader.GetString 2
    }

  let rec getSeqFromDataReader (dataReader:SqlDataReader) : StudentCode list =
    match dataReader.Read() with
    | true  -> getStudentCodeFromDataReader( dataReader ) :: getSeqFromDataReader(dataReader)
    | false -> []

  override this.Get id =    
    Async.StartAsTask( async {
      use connection = new SqlConnection( connectionString )
      connection.Open()

      use command = new SqlCommand( "SELECT Id, Email, Code FROM StudentCodes WHERE Id = @Id", connection )
      command.Parameters.Add( "Id", SqlDbType.Int ).Value <- id

      use! dataReader = command.ExecuteReaderAsync() |> Async.AwaitTask
      match dataReader.Read() with
      | true  -> return Some {
        Id    = dataReader.GetInt32 0;
        Email = dataReader.GetString 1;
        Code  = dataReader.GetString 2 }
      | false -> return None
    } )

  override this.Get( field, value ) =
    Async.StartAsTask( async {
      use connection = new SqlConnection( connectionString )
      connection.Open()

      use command = new SqlCommand( "SELECT Id, Email, Code FROM StudentCodes WHERE " + field + " = @value", connection )
      command.Parameters.AddWithValue( "value", value ) |> ignore

      use! dataReader = command.ExecuteReaderAsync() |> Async.AwaitTask
      return getSeqFromDataReader dataReader
    } )

  override this.Add itemToAdd =
    Async.StartAsTask( async {
      use connection = new SqlConnection( connectionString )
      connection.Open()

      use command = new SqlCommand( "INSERT StudentCodes ( Email, Code ) VALUES ( @Email, @Code ); SELECT SCOPE_IDENTITY();", connection )
      command.Parameters.Add( "Email", SqlDbType.VarChar ).Value <- itemToAdd.Email
      command.Parameters.Add( "Code", SqlDbType.VarChar ).Value <- itemToAdd.Code

      let! id = command.ExecuteScalarAsync() |> Async.AwaitTask
      return { itemToAdd with Id = System.Convert.ToInt32( id ) }
    } )
